### ISSUE TITLE
Probably Bugs with the NIFD converter

### ISSUE THREAD

**Issue by Kateridge** [aramis-lab/clinica#929]
**Describe the bug**
There are some errors when I use the converter.
1) Error at nifd_utils.py line 247 
```python
session_id=lambda df: df.visit.apply(lambda x: f"ses-M{(6 * (x - 1)):03d}"),
```
The visit format for the collection data is 'M0', 'M6', ..., which is different from the clinical data file (use a single number). I modify the code into:
```python
session_id=lambda df: df.visit.apply(lambda x: "ses-M" + x.strip('M').zfill(3)),
```

2) Error at nifd_utils.py line 273-279
```python
sessions = (
    scans[["participant_id", "session_id", "acq_date", "age"]]
    .rename(columns={"acq_date": "date"})
    .drop_duplicates()
    .set_index(["participant_id", "session_id"], verify_integrity=True)
    .sort_index()
).join(clinical_data[["cdr", "mmse"]])
```
If we have multi-tracer PET in the same session, there will be duplicate index before .set_index(...). We have to drop duplicate sessions according to participant_id and session_id. I modify the line 276 into:
```python
.drop_duplicates(subset=["participant_id", "session_id"])
```

3) The converter only convert T1 and FDG-PET scans. There are no PIB-PET scans in the BIDS directory after running the converter. The problem should happen at nifd_utils.py line 301-303:
```python
if fs.exists(output_dir):
    fs.rm(output_dir, recursive=True)
fs.makedirs(output_dir)
```
If we have multi-tracer PET scans at the same session, the above code will delete the whole sub-\*/ses-M\*/pet/ folder. The previous converted PIB-PET with be deleted when we convert the FDG-PET in the same session. The code could be modified as follows:
```python
if not fs.exists(output_dir):
    fs.makedirs(output_dir)
```

**To Reproduce**
Just follow the instruction to convert NIFD dataset to BIDS format. I put clinical data 'NIFD_Clinical_Data_20200724_updated.xlsx' and 'DataDictionary_NIFD_2017.10.18.xlsx' in the clinical data directory. And I also put the collection metadata 'idaSearch_all.csv' in both clinical data and original dataset directory.

**Expected behavior**
Convert all modalities successfully (T1, FDG-PET, PIB-PET in my case)

**Comment by ghisvail** [comment]
Hi @Kateridge,

> The visit format for the collection data is 'M0', 'M6', ..., which is different from the clinical data file (use a single number).

This is a convention used by Clinica for all its converters. There is a normalisation step happening to convert the session coding used by the source dataset to the `M{:03d}` coding scheme used within Clinica.

> If we have multi-tracer PET in the same session, there will be duplicate index before .set_index(...)

You are correct. Could you provide a pull-request with your proposed fix please?

> The converter only convert T1 and FDG-PET scans.

Same as above. A separate PR with your proposed fix would be very much appreciated 🙏 

Thank you very much for taking the time to formulate such a clear report of your issues 👍

**Comment by Kateridge** [comment]
> This is a convention used by Clinica for all its converters. There is a normalisation step happening to convert the session coding used by the source dataset to the M{:03d} coding scheme used within Clinica.

Sorry for the unclear description about the first issue. I understand Clinica is going to convert the session coding into a uniform format M{:03d}. Here, the problem is that, visit coding format of NIFD collection data (idaSearch_all.csv) is 'M0', 'M6', 'M12',... . The code at nifd_utils.py line 247 `session_id=lambda df: df.visit.apply(lambda x: f"ses-M{(6 * (x - 1)):03d}")` will throw an error like 'str cannot be subtracted by an int', because x here is a str 'M*'.

![QQ截图20230528182554](https://github.com/aramis-lab/clinica/assets/29402665/33306d7f-dcf7-410d-9132-f7fbc00c349b)

Besides, I find another problem when I run pet-linear pipeline on the converted dataset. The pet-linear pipeline cannot find any PET file. This problem probably comes from the naming format of PET files. The NIFD converter uses `sub-NIFD*_ses-M*_trc-*_rec-*_pet.nii.gz` to name the converted PET files, which cannot be matched by the filename pattern `*_trc-{tracer}_pet.nii*` used by pet-linear pipeline. Moving "rec-\*" in front of "trc-\*" could solve this problem.

If above-mentioned content is correct, I will also include these fixes in the PR.

**Comment by ghisvail** [comment]
I am glad you provided the screenshot 😅 

I looked at my test data from a collection and clinical data I downloaded back in 2021. The visit encoding changed from integers (1 -> baseline, 2 -> 6-month, 3 -> 12-month...) to a string in `MXX` format 🤦 

Indeed, that piece of logic needs to be changed to accomodate with the new encoding. I'll check with the rest of the team whether we want to maintain backward compatibility with the old encoding. Meanwhile, feel free to provide a PR for it with the code that works for your use case and I'll make sure you'll be appropriately credited in the final solution.

**Comment by ghisvail** [comment]
> Besides, I find another problem when I run pet-linear pipeline on the converted dataset. The pet-linear pipeline cannot find any PET file. 

Indeed, that's a separate issue with our PET pipelines. Historically, they were designed with mono-tracer PET datasets in mind and the regexes used inside them aren't flexible enough to accomodate with additional entities such as multiple runs or tracers. We are aware of this issue, but it's good to have a reminder that other people in the community care for it.

Once again, feel free to submit a PR with what works for you.

> Moving "rec-*" in front of "trc-*" could solve this problem.

This is not an option as it would make the dataset no longer BIDS compliant. The entities must be ordered in a [certain way](https://bids-specification.readthedocs.io/en/stable/appendices/entity-table.html#positron-emission-tomography-pet).

**Comment by ghisvail** [comment]
> If above-mentioned content is correct, I will also include these fixes in the PR.

We would prefer to have fixes concerning the NIFD converter and the PET pipelines in separate PRs :pray:

**Comment by ghisvail** [comment]
Just a heads-up. I have included your fix in #932 with appropriate co-authorship.

We will treat the issue with file parsing in PET pipelines in a different PR.

Once again, thank you very much for investigating this issue and providing a solution 👍
